cuBERT_topic_modelling/mmr.py

```diff
@@ -0,0 +1,61 @@
+from typing import List
+import cupy as cp
+from cuml.metrics import pairwise_distances
+
+
+def mmr(
+    doc_embedding: cp.ndarray,
+    word_embeddings: cp.ndarray,
+    words: List[str],
+    top_n: int = 5,
+    diversity: float = 0.8,
+):
+
+    """
+    Calculate Maximal Marginal Relevance (MMR)
+    between candidate keywords and the document.
+    MMR considers the similarity of keywords/keyphrases with the
+    document, along with the similarity of already selected
+    keywords and keyphrases. This results in a selection of keywords
+    that maximize their within diversity with respect to the document.
+    Arguments:
+        doc_embedding: The document embeddings
+        word_embeddings: The embeddings of the selected candidate keywords/phrases
+        words: The selected candidate keywords/keyphrases
+        top_n: The number of keywords/keyhprases to return
+        diversity: How diverse the select keywords/keyphrases are.
+                   Values between 0 and 1 with 0 being not diverse at all
+                   and 1 being most diverse.
+    Returns:
+         List[str]: The selected keywords/keyphrases
+    """
+
+    # Extract similarity within words, and between words and the document
+    word_doc_similarity = pairwise_distances(
+        word_embeddings, doc_embedding, metric="cosine"
+    )
+    word_similarity = pairwise_distances(word_embeddings, metric="cosine")
+
+    # Initialize candidates and already choose best keyword/keyphras
+    keywords_idx = [cp.argmax(word_doc_similarity).get()]
```

Comment:
Why did we convert to numpy here ?

Reply:
Does `cupy.take` not work ? 

 https://docs.cupy.dev/en/stable/reference/generated/cupy.take.html#cupy.take
 
I suggest looking more into this, very surprised that indexing a cupy array with cupy is not allowed.

Reply:
yeah I just came across this as well, I'll look into it, thanks!